### DISCUSSION TITLE
Added a new feature

### DISCUSSION THREAD

**Discussion by kdschlosser** [lvgl-micropython/lvgl_micropython#129] [Announcements]
I added a feature to the touch drivers that allows user input to have the highest priority. So when a user does something like scrolling the display it should make the scrolling a lot smoother. It will also update the display immediately if there is user input instead of waiting for the refresh timer to update the display.

To use this feature call the method `enable_input_priority` in the touch driver...

```
indev = gt911,GT911(...)
indev.enable_input_priority()
```

it's as simple as that...

**Comment by fabse-hack** [comment]
nice! i have to order more displays 😆 

thank you for all your work! is amazing & game changing for micropython and all projects with a display.

**Reply by kdschlosser** [reply]
I am trying to make it better as I go along.  When I see deficiencies somewhere I am doing my best to provide a way to work around it. This specific feature I made it optional because of how it works. It makes user input have more priority than anything else. It stalls the main loop while the user is interacting with the device. More often than not a lot of user code is actually placed inside of event callbacks for different UI elements. Those callbacks still get processed. the only thing that doesn't is the code in the main loop if the user has code in the main loop. 

My recommendation is always to not have code in the main loop because this will tie up the REPL. It is always better to spawn a thread to run any looping code in. I did add another feature but it has not been tested as of yet. It allows threads to be created on the second core. In order for this to happen the GIL had to be removed so things can actually run in parallel to each other. Without the GIL the user has to be super careful about guarding parallel access to things. I haven't mentions that this feature exists yet because I wanted to expose some of the FreeRTOS mechanics so instead of just having locks available there would also be events to create stalls that are not wheel spinners and queues to pass things between the threads safely.

**Reply by fabse-hack** [reply]
i defenitly used a translator, to understand you answer 100%. I hope in the future in will learn about it more! ....

Yes, i have the same problem, i take code in the main look. Asyncio is implementated, for example like this:

```
async def main():
    task1 = asyncio.create_task(data_to_lvgl_every_hours())
    task2 = asyncio.create_task(data_to_lvgl_every_second())
    task3 = asyncio.create_task(mytime.tick())
    task4 = asyncio.create_task(mqtt_publisher.run())
    await asyncio.gather(task1, task2, task3, task4)
```

it's working .... but not the goal.

**Reply by kdschlosser** [reply]
don't bother with asynchio. most of it runs in Python code in MicroPython. it is slow and doesn't offer any real benefit.

**Reply by dcmcshan** [reply]
Very interested in running on Core 0 - do you have test code I can try?

**Reply by kdschlosser** [reply]
everything is always running on core 0. core 1 is what doesn't get used by MicroPython. core 1 is what gets used for Blueteeth and Wifi and that's about all it gets used for I think.

**Reply by fabse-hack** [reply]
> don't bother with asynchio. most of it runs in Python code in MicroPython. it is slow and doesn't offer any real benefit.

hmmm but, i have some simultan runs, how i can handle it without asyncio ? i think a good example is a blinky led, how the led can blink and call some other functions on the same time? ...... (okay, maybe can code with some spagetthi code, but not with object orientated programming) ...or with threads?

**Reply by kdschlosser** [reply]
threads!!! you got it....

When using MicroPython your code should run in a loop. The "main thread" is the REPL and it should be left to handle that. all of your code that runs should be running in a different thread that runs in a loop. This will allow you to connect via serial and get a REPL prompt. You should add an exit routing for the thread that you can run from the REPL prompt that will tell the thread to exit.

I would need to see your code in order to determine the best plan of action on how to set things up.

**Reply by fabse-hack** [reply]
ah thanks!!! I will look at this. I just found out, that threads is available  🤓

**Comment by picpic020960** [comment]
**AIOREPPL** is good for main loop control, no?
